README
NAME
    Dancer::Session::Cookie - Encrypted cookie-based session backend for
    Dancer

VERSION
    version 0.25

SYNOPSIS
    Your config.yml:

        session: "cookie"
        session_cookie_key: "this random key IS NOT very random"

DESCRIPTION
    This module implements a session engine for sessions stored entirely in
    cookies. Usually only session id is stored in cookies and the session
    data itself is saved in some external storage, e.g. database. This
    module allows to avoid using external storage at all.

    Since server cannot trust any data returned by client in cookies, this
    module uses cryptography to ensure integrity and also secrecy. The data
    your application stores in sessions is completely protected from both
    tampering and analysis on the client-side.

CONFIGURATION
    The setting session should be set to "cookie" in order to use this
    session engine in a Dancer application. See Dancer::Config.

    A mandatory setting is needed as well: session_cookie_key, which should
    contain a random string of at least 16 characters (shorter keys are not
    cryptographically strong using AES in CBC mode).

    Here is an example configuration to use in your config.yml:

        session: "cookie"
        session_cookie_key: "kjsdf07234hjf0sdkflj12*&(@*jk"

    Compromising session_cookie_key will disclose session data to clients
    and proxies or eavesdroppers and will also allow tampering, for example
    session theft. So, your config.yml should be kept at least as secure as
    your database passwords or even more.

    Also, changing session_cookie_key will have an effect of immediate
    invalidation of all sessions issued with the old value of key.

    session_cookie_path can be used to control the path of the session
    cookie. The default is /.

    The global session_secure setting is honoured and a secure (https only)
    cookie will be used if set.

DEPENDENCY
    This module depends on Session::Storage::Secure. Legacy support is
    provided using Crypt::CBC, Crypt::Rijndael, String::CRC32, Storable and
    MIME::Base64.

SEE ALSO
    See Dancer::Session for details about session usage in route handlers.

    See Plack::Middleware::Session::Cookie,
    Catalyst::Plugin::CookiedSession, "session" in Mojolicious::Controller
    for alternative implementation of this mechanism.

AUTHORS
    *   Alex Kapranoff <kappa@cpan.org>

    *   Alex Sukria <sukria@cpan.org>

    *   David Golden <dagolden@cpan.org>

    *   Yanick Champoux <yanick@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Alex Kapranoff.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Dancer-Session-Cookie

{{$NEXT}}
  [ API CHANGES ]

  [ BUG FIXES ]

  [ DOCUMENTATION ]

  [ ENHANCEMENTS ]

  [ NEW FEATURES ]

  [ STATISTICS ]

0.30 2018-03-11
  [ ENHANCEMENTS ]
    - add a CONTRIBUTING file [GH#18, Paul Cochrane]
    - minor refactoring, bump minor Perl version to 5.10.0.

  [ STATISTICS ]
    - code churn: 5 files changed, 127 insertions(+), 26 deletions(-)

0.29 2018-02-19
  [ MISC ]
    - Added appveyor integration for Windows CI. [GH#11, Paul Cochrane]
    - Set minimal Perl version to be 5.8. [GH#10, Paul Cochrane]
    - Misc code cleanup. [GH#14, GH#15, GH#16, GH#17, Paul Cochrane]
    - Added appveyor and travis badges to README. [GH#13, Paul Cochrane]

  [ STATISTICS ]
    - code churn: 12 files changed, 69 insertions(+), 21 deletions(-)

0.28 2018-02-10
  [ BUG FIXES ]
    - Tests require Plack::Test v1.0029+. [GH#8, Mohammad S Anwar]

  [ DOCUMENTATION ]
    - Improve English flow of the POD. [GH#9, Paul Cochrane]

  [ STATISTICS ]
    - code churn: 10 files changed, 105 insertions(+), 48 deletions(-)

0.27 2015-09-23
  - Remode  'README.pod' from distribution. [GH#6]

  [ STATISTICS ]
    - code churn: 4 files changed, 62 insertions(+), 126 deletions(-)

0.26 2015-09-10
  - Switch Test::WWW::Mechanize::PSGI for raw HTTP::Request/Response.
    (GH#5, xsawyerx)

  [ DOCUMENTATION ]
    - Document the 'session_expires' setting. - Warn about maximal cookie
      size. (GH#2 David Precious)

  [ STATISTICS ]
    - code churn: 10 files changed, 304 insertions(+), 192 deletions(-)

0.25 2014-08-04
  [ MISC ]
    - Move tests from Test::TCP to Test::WWW::Mechanize::PSGI.

  [ STATISTICS ]
    - code churn: 5 files changed, 138 insertions(+), 174 deletions(-)

0.24 2014-07-29
  [ MISC ]
    - Release again, this time with real co-maint permissions.

  [ STATISTICS ]
    - code churn: 1 file changed, 57 insertions(+), 53 deletions(-)

0.23 2014-07-17
  [ CHANGED ]
    - Requires Session::Storage::Secure 0.010 to allow storing objects,
      which is specially relevant for JSON::bool data.

  [ STATISTICS ]
    - code churn: 8 files changed, 67 insertions(+), 119 deletions(-)

0.22 2013-05-31T23:38:07Z America/New_York
  [ CHANGED ]
    - Requires Session::Storage::Secure 0.007 for improved resistance to
      timing attacks

0.21 2013-05-08T22:33:10Z America/New_York
  - Require Dancer 1.3113 and use its new features to fix more bugs that
    caused sessions to not be set or reset correctly

0.20 2013-04-25T11:41:22Z America/New_York
  - Fix bug where session wasn't cleared correctly if route handler died
    while processing a request

0.19 2013-02-24T22:49:14Z America/New_York
  - Fix bug with session_expires always expiring a cookie early

0.18 2013-02-16T00:16:32Z America/New_York
  - Hack around Dancer bug that renders the request object headers
    repeatedly and locks out setting cookies after the first time

0.17 2013-02-07T15:36:58Z America/New_York
  - Fix localhost address used in tests

0.16 2013-02-01T14:39:14Z America/New_York
  - Now uses Session::Storage::Secure for encrypting and decrypting data;
    Legacy decryption is preserved in case old cookies need to be read.
  - Fix session destruction bug to ensure session is overwritten in the
    browser, not just no longer sent (David Golden)

0.15 2011-05-21
  - skip test 02-configfile.t if YAML is not present FIX smoker test
    failure. (Alexis Sukrieh)

0.14 2011-03-12
  [ Michael G. Schwern ]
    - Fix for Dancer > v1.3012
    - Add support for session_secure to serve https only cookies.
    - Add missing MYMETA.yml
    - Make Dancer::Session::Cookie honor the session_name setting added to
      Dancer::Session::Abstract
    - Add session_cookie_path to control the path of the cookie.

0.13 2010-11-28T17:19:58Z
  - Some documentation fixes. No functional change.

0.12 2010-09-16T13:29:09Z
  - Fix a warning when testing against newer Dancer. No functional
  - change.

0.11 2010-02-17T16:12:21Z
  - Chase the ever-changing Dancer core :) No functional change.
  - More tests.

0.1 2010-02-02T17:20:11Z
  - First version separated from the Dancer core. No functional
  - change.


Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.019.
use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601"
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Encrypted cookie-based session backend for Dancer",
  "dist_author" => [
    "Alex Kapranoff <kappa\@cpan.org>",
    "Alex Sukria <sukria\@cpan.org>",
    "David Golden <dagolden\@cpan.org>",
    "Yanick Champoux <yanick\@cpan.org>"
  ],
  "dist_name" => "Dancer-Session-Cookie",
  "dist_version" => "0.25",
  "license" => "perl",
  "module_name" => "Dancer::Session::Cookie",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Crypt::CBC" => 0,
    "Crypt::Rijndael" => 0,
    "Dancer" => "1.3113",
    "Dancer::Cookie" => 0,
    "Dancer::Cookies" => 0,
    "Dancer::Session::Abstract" => 0,
    "MIME::Base64" => 0,
    "Session::Storage::Secure" => "0.010",
    "Storable" => 0,
    "String::CRC32" => 0,
    "Time::Duration::Parse" => 0,
    "base" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "Dancer::ModuleLoader" => 0,
    "Dancer::Test" => 0,
    "File::Spec" => 0,
    "File::Temp" => 0,
    "FindBin" => 0,
    "HTTP::Date" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::Exception" => 0,
    "Test::More" => "0.96",
    "Test::NoWarnings" => 0,
    "Test::WWW::Mechanize::PSGI" => 0,
    "perl" => "5.006"
  }
);


my %fallback_build_requires = (
  "Dancer::ModuleLoader" => 0,
  "Dancer::Test" => 0,
  "File::Spec" => 0,
  "File::Temp" => 0,
  "FindBin" => 0,
  "HTTP::Date" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Module::Build" => "0.3601",
  "Test::Exception" => 0,
  "Test::More" => "0.96",
  "Test::NoWarnings" => 0,
  "Test::WWW::Mechanize::PSGI" => 0,
  "perl" => "5.006"
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/Dancer/Session/Cookie.pm
package Dancer::Session::Cookie;
# ABSTRACT: Encrypted cookie-based session backend for Dancer

use strict;
use warnings;

use 5.10.0;
use parent 'Dancer::Session::Abstract';

use Session::Storage::Secure 0.010;
use Crypt::CBC;
use String::CRC32;
use Crypt::Rijndael;
use Time::Duration::Parse;

use Dancer 1.3113 ':syntax'; # 1.3113 for on_reset_state and fixed after hook
use Dancer::Cookie  ();
use Dancer::Cookies ();
use Storable        ();
use MIME::Base64    ();
use PerlX::Maybe;

# crydec
my $CIPHER = undef;
my $STORE  = undef;

# cache session here instead of flushing/reading from cookie all the time
my $SESSION = undef;

sub is_lazy { 1 }; # avoid calling flush needlessly

sub init {
    my ($self) = @_;

    $self->SUPER::init();

    my $key = setting("session_cookie_key") # XXX default to smth with warning
      or die "The setting session_cookie_key must be defined";

    my $duration = $self->_session_expires_as_duration;

    $CIPHER = Crypt::CBC->new(
        -key    => $key,
        -cipher => 'Rijndael',
    );

    $STORE = Session::Storage::Secure->new(
        secret_key => $key,
        sereal_encoder_options => { snappy => 1, stringify_unknown => 1 },
        sereal_decoder_options => { validate_utf8 => 1 },
        maybe default_duration => $duration,
    );
}

# return our cached ID if we have it instead of looking in a cookie
sub read_session_id {
    my ($self) = @_;
    return defined $SESSION 
                ? $SESSION->id
                : $self->SUPER::read_session_id;
}

sub retrieve {
    my ( $class, $id ) = @_;
    # if we have a cached session, hand that back instead
    # of decrypting again
    return $SESSION
      if $SESSION && $SESSION->id eq $id;

    my $ses = eval {
        if ( my $hash = $STORE->decode($id) ) {
            # we recover a plain hash, so reconstruct into object
            bless $hash, $class;
        }
        else {
            _old_retrieve($id);
        }
    };

    return $SESSION = $ses;
}

# support decoding old cookies
sub _old_retrieve {
    my ($id) = @_;
    # 1. decrypt and deserialize $id
    my $plain_text = _old_decrypt($id);
    # 2. deserialize
    $plain_text && Storable::thaw($plain_text);
}

sub create {
    # cache the newly created session
    return $SESSION = Dancer::Session::Cookie->new;
}

# we don't write session ID when told; we do it in the after hook
sub write_session_id { }

# we don't flush when we're told; we do it in the after hook
sub flush { }

sub destroy {
    my $self = shift;

    # gross hack; replace guts with new session guts
    %$self = %{ Dancer::Session::Cookie->new };

    return 1;
}

# Copied from Dancer::Session::Abstract::write_session_id and
# refactored for testing
hook 'after' => sub {
    my $response = shift;

    return unless $SESSION;

    # UGH! Awful hack because Dancer instantiates responses
    # and headers too many times and locks out new cookies
    $response->{_built_cookies} = 0;

    my $c = Dancer::Cookie->new( $SESSION->_cookie_params );
    Dancer::Cookies->set_cookie_object( $c->name => $c );
};

# Make sure that the session is initially undefined for every request
hook 'on_reset_state' => sub {
    my $is_forward = shift;
    undef $SESSION unless $is_forward;
};

# modified from Dancer::Session::Abstract::write_session_id to add
# support for session_cookie_path
sub _cookie_params {
    my $self     = shift;
    my $name     = $self->session_name;
    my $duration = $self->_session_expires_as_duration;
    my %cookie   = (
        name      => $name,
        value     => $self->_cookie_value,
        path      => setting('session_cookie_path') || '/',
        domain    => setting('session_domain'),
        secure    => setting('session_secure'),
        http_only => setting("session_is_http_only") // 1.
    );
    if ( defined $duration ) {
        $cookie{expires} = time + $duration;
    }
    return %cookie;
}

# refactored for testing
sub _cookie_value {
    my ($self) = @_;
    # copy self guts so we aren't serializing a blessed object.
    # we don't set expires, because default_duration will handle it
    return $STORE->encode( {%$self} );
}

# session_expires could be natural language
sub _session_expires_as_duration {
    my ($self) = @_;
    my $session_expires = setting('session_expires');
    return unless defined $session_expires;
    my $duration = eval { parse_duration($session_expires) };
    die "Could not parse session_expires: $session_expires"
      unless defined $duration;
    return $duration;
}

# legacy algorithm
sub _old_decrypt {
    my $cookie = shift;

    $cookie =~ tr{_*-}{=+/};

    $SIG{__WARN__} = sub { };
    my ( $crc32, $plain_text ) = unpack "La*",
      $CIPHER->decrypt( MIME::Base64::decode($cookie) );
    return $crc32 == String::CRC32::crc32($plain_text) ? $plain_text : undef;
}

1;
__END__

=head1 SYNOPSIS

Your F<config.yml>:

    session: "cookie"
    session_cookie_key: "this random key IS NOT very random"

=head1 DESCRIPTION

This module implements a session engine for sessions stored entirely
in cookies. Usually only the B<session id> is stored in cookies and
the session data itself is saved in some external storage, e.g.
a database. This module allows you to avoid using external storage at
all.

Since a server cannot trust any data returned by clients in cookies, this
module uses cryptography to ensure integrity and also secrecy. The
data your application stores in sessions is completely protected from
both tampering and analysis on the client-side.

Do be aware that browsers limit the size of individual cookies, so this method
is not suitable if you wish to store a large amount of data.  Browsers typically
limit the size of a cookie to 4KB, but that includes the space taken to store
the cookie's name, expiration and other attributes as well as its content.

=head1 CONFIGURATION

The setting B<session> should be set to C<cookie> in order to use this session
engine in a Dancer application. See L<Dancer::Config>.

Another setting is also required: B<session_cookie_key>, which should
contain a random string of at least 16 characters (shorter keys are
not cryptographically strong using AES in CBC mode).

The optional B<session_expires> setting can also be passed,
which will provide the duration time of the cookie. If it's not present, the
cookie won't have an expiration value.

Here is an example configuration to use in your F<config.yml>:

    session: "cookie"
    session_cookie_key: "kjsdf07234hjf0sdkflj12*&(@*jk"
    session_expires: 1 hour

Compromising B<session_cookie_key> will disclose session data to
clients and proxies or eavesdroppers and will also allow tampering,
for example session theft. So, your F<config.yml> should be kept at
least as secure as your database passwords or even more.

Also, changing B<session_cookie_key> will have an effect of immediate
invalidation of all sessions issued with the old value of key.

B<session_cookie_path> can be used to control the path of the session
cookie.  The default is C</>.

The global B<session_secure> setting is honored and a secure (https
only) cookie will be used if set.

=head1 DEPENDENCY

This module depends on L<Session::Storage::Secure>.  Legacy support is provided
using L<Crypt::CBC>, L<Crypt::Rijndael>, L<String::CRC32>, L<Storable> and
L<MIME::Base64>.

=head1 SEE ALSO

See L<Dancer::Session> for details about session usage in route handlers.

See L<Plack::Middleware::Session::Cookie>,
L<Catalyst::Plugin::CookiedSession>, L<Mojolicious::Controller/session> for alternative implementation of this mechanism.

=cut

# vim: ts=4 sts=4 sw=4 et:


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.009.
use strict;
use warnings;

use 5.008;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Encrypted cookie-based session backend for Dancer",
  "AUTHOR" => "Alex Kapranoff <kappa\@cpan.org>, Alex Sukria <sukria\@cpan.org>, David Golden <dagolden\@cpan.org>, Yanick Champoux <yanick\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Dancer-Session-Cookie",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.008",
  "NAME" => "Dancer::Session::Cookie",
  "PREREQ_PM" => {
    "Crypt::CBC" => 0,
    "Crypt::Rijndael" => 0,
    "Dancer" => "1.3113",
    "Dancer::Cookie" => 0,
    "Dancer::Cookies" => 0,
    "Dancer::Session::Abstract" => 0,
    "MIME::Base64" => 0,
    "Session::Storage::Secure" => "0.010",
    "Storable" => 0,
    "String::CRC32" => 0,
    "Time::Duration::Parse" => 0,
    "base" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Dancer::ModuleLoader" => 0,
    "Dancer::Test" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "File::Temp" => 0,
    "FindBin" => 0,
    "HTTP::Cookies" => 0,
    "HTTP::Date" => 0,
    "HTTP::Request::Common" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Plack" => "1.0029",
    "Plack::Test" => 0,
    "Test::Exception" => 0,
    "Test::More" => "0.96",
    "Test::NoWarnings" => 0,
    "Test::Requires" => 0
  },
  "VERSION" => "0.29",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Crypt::CBC" => 0,
  "Crypt::Rijndael" => 0,
  "Dancer" => "1.3113",
  "Dancer::Cookie" => 0,
  "Dancer::Cookies" => 0,
  "Dancer::ModuleLoader" => 0,
  "Dancer::Session::Abstract" => 0,
  "Dancer::Test" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "File::Temp" => 0,
  "FindBin" => 0,
  "HTTP::Cookies" => 0,
  "HTTP::Date" => 0,
  "HTTP::Request::Common" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "MIME::Base64" => 0,
  "Plack" => "1.0029",
  "Plack::Test" => 0,
  "Session::Storage::Secure" => "0.010",
  "Storable" => 0,
  "String::CRC32" => 0,
  "Test::Exception" => 0,
  "Test::More" => "0.96",
  "Test::NoWarnings" => 0,
  "Test::Requires" => 0,
  "Time::Duration::Parse" => 0,
  "base" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
